File: skills/omnifocus-manager/scripts/generate_plugin.py
#!/usr/bin/env python3
"""
OmniFocus Plugin Generator

Quickly create plugins from templates with correct API patterns.
Supports both solitary (single-file) and bundle (folder) formats.

Usage:
    # Solitary action plugin (simple, single file)
    python3 generate_plugin.py --format solitary --name "My Quick Action"

    # Solitary action with Foundation Models
    python3 generate_plugin.py --format solitary-fm --name "AI Analyzer"

    # Solitary library plugin
    python3 generate_plugin.py --format solitary-library --name "My Utils"

    # Bundle plugin (multiple actions, libraries, localization)
    python3 generate_plugin.py --format bundle --template query-simple --name "My Plugin"
"""

import argparse
import os
import shutil
import subprocess
import sys
from pathlib import Path

# Plugin formats
FORMATS = {
    'solitary': 'Single-file action plugin (.omnijs)',
    'solitary-fm': 'Single-file action plugin with Foundation Models (.omnijs)',
    'solitary-library': 'Single-file library plugin (.omnijs)',
    'bundle': 'Bundle plugin with localization (.omnifocusjs folder)',
}

# Bundle templates
BUNDLE_TEMPLATES = {
    'query-simple': 'Simple query plugin with Alert display',
    'stats-overview': 'Statistics overview dashboard with detailed breakdowns',
}

TEMPLATE_VARS = {
    'PLUGIN_NAME': 'Plugin display name',
    'IDENTIFIER': 'Reverse domain identifier (e.g., com.user.myplugin)',
    'AUTHOR': 'Plugin author name',
    'DESCRIPTION': 'Plugin description',
    'ACTION_ID': 'Action identifier (camelCase)',
    'ACTION_LABEL': 'Action label shown in menus',
    'SHORT_LABEL': 'Short label (1-2 words)',
    'MEDIUM_LABEL': 'Medium label',
    'LONG_LABEL': 'Long descriptive label',
    'PALETTE_LABEL': 'Palette/toolbar label',
    'ICON': 'SF Symbol icon name (e.g., list.bullet, chart.bar)',
}


def find_omnifocus_manager_root():
    """Find the omnifocus-manager skill root directory."""
    current = Path(__file__).resolve().parent
    while current != current.parent:
        if (current / 'SKILL.md').exists() and current.name == 'omnifocus-manager':
            return current
        current = current.parent
    return None


def replace_template_vars(content, variables):
    """Replace template variables in content."""
    for key, value in variables.items():
        placeholder = '{{' + key + '}}'
        content = content.replace(placeholder, value)
    return content


def process_file(file_path, variables):
    """Process a single file, replacing template variables."""
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()

        content = replace_template_vars(content, variables)

        with open(file_path, 'w', encoding='utf-8') as f:
            f.write(content)
        return True
    except Exception as e:
        print(f"  ⚠️  Error processing {file_path}: {e}")
        return False


def camel_case(text):
    """Convert text to camelCase."""
    words = text.replace('-', ' ').replace('_', ' ').split()
    if not words:
        return ''
    return words[0].lower() + ''.join(word.capitalize() for word in words[1:])


def validate_javascript(file_path):
    """Validate JavaScript file using eslint_d."""
    try:
        result = subprocess.run(
            ['eslint_d', str(file_path)],
            capture_output=True,
            text=True,
            timeout=10
        )

        if result.returncode == 0:
            print(f"  ✅ {file_path.name} - no linting errors")
            return True
        else:
            print(f"  ⚠️  {file_path.name} - linting issues found:")
            print(result.stdout)
            return False
    except subprocess.TimeoutExpired:
        print(f"  ⚠️  {file_path.name} - validation timeout")
        return False
    except FileNotFoundError:
        print(f"  ℹ️  eslint_d not found - skipping validation")
        return True
    except Exception as e:
        print(f"  ⚠️  {file_path.name} - validation error: {e}")
        return False


def generate_solitary_plugin(format_type, name, identifier=None, author=None, description=None, output_dir=None):
    """Generate a solitary (single-file) plugin."""

    root = find_omnifocus_manager_root()
    if not root:
        print("❌ Error: Could not find omnifocus-manager root directory")
        return False

    # Set defaults
    if not identifier:
        identifier = f"com.user.omnifocus.{name.lower().replace(' ', '')}"
    if not author:
        author = "OmniFocus User"
    if not description:
        description = f"{name} plugin for OmniFocus"
    if not output_dir:
        output_dir = root / 'assets'

    # Determine template based on format
    if format_type == 'solitary':
        template_file = root / 'assets' / 'plugin-templates' / 'solitary-action' / 'action.omnijs.template'
        output_ext = '.omnijs'
    elif format_type == 'solitary-fm':
        template_file = root / 'assets' / 'plugin-templates' / 'solitary-action-fm' / 'action.omnijs.template'
        output_ext = '.omnijs'
    elif format_type == 'solitary-library':
        template_file = root / 'assets' / 'plugin-templates' / 'solitary-library' / 'library.omnijs.template'
        output_ext = '.omnijs'
    else:
        print(f"❌ Error: Unknown solitary format '{format_type}'")
        return False

    if not template_file.exists():
        print(f"❌ Error: Template not found at {template_file}")
        return False

    # Output path
    output_filename = f"{name.replace(' ', '')}{output_ext}"
    output_path = Path(output_dir) / output_filename

    if output_path.exists():
        print(f"❌ Error: Plugin already exists at {output_path}")
        response = input("Overwrite? (yes/no): ")
        if response.lower() not in ['yes', 'y']:
            print("Cancelled.")
            return False
        output_path.unlink()

    # Prepare template variables
    action_id = camel_case(name)
    short_label = name.split()[0] if ' ' in name else name
    variables = {
        'PLUGIN_NAME': name,
        'IDENTIFIER': identifier,
        'AUTHOR': author,
        'DESCRIPTION': description,
        'ACTION_ID': action_id,
        'ACTION_LABEL': name,
        'SHORT_LABEL': short_label,
        'MEDIUM_LABEL': name,
        'LONG_LABEL': f"{name} with OmniFocus Automation",
        'PALETTE_LABEL': name,
        'ICON': 'apple.intelligence' if format_type == 'solitary-fm' else 'list.bullet',
    }

    print(f"\n🔨 Generating solitary plugin...")
    print(f"   Format: {format_type}")
    print(f"   Plugin: {name}")
    print(f"   Identifier: {identifier}")
    print(f"   Output: {output_path}\n")

    # Read and process template
    try:
        with open(template_file, 'r', encoding='utf-8') as f:
            content = f.read()

        content = replace_template_vars(content, variables)

        with open(output_path, 'w', encoding='utf-8') as f:
            f.write(content)

        print(f"✅ Created {output_path}")
    except Exception as e:
        print(f"❌ Error creating plugin: {e}")
        return False

    # Validate
    print(f"\n🔍 Validating JavaScript...")
    validate_javascript(output_path)

    # Success message
    print(f"\n🎉 Plugin generated successfully!")
    print(f"\n📦 Installation:")
    print(f"   cp {output_path} ~/Library/Application\\ Scripts/com.omnigroup.OmniFocus3/Plug-Ins/")
    print(f"\n🧪 Testing:")
    print(f"   1. Restart OmniFocus (if running)")
    print(f"   2. Go to Automation menu")
    print(f"   3. Find '{name}'")

    return True


def generate_bundle_plugin(template, name, identifier=None, author=None, description=None, output_dir=None):
    """Generate a bundle plugin from a template."""

    root = find_omnifocus_manager_root()
    if not root:
        print("❌ Error: Could not find omnifocus-manager root directory")
        return False

    # Set defaults
    if not identifier:
        identifier = f"com.user.omnifocus.{name.lower().replace(' ', '')}"
    if not author:
        author = "OmniFocus User"
    if not description:
        description = f"{name} plugin for OmniFocus"
    if not output_dir:
        output_dir = root / 'assets'

    # Template paths
    template_dir = root / 'assets' / 'plugin-templates' / template
    if not template_dir.exists():
        print(f"❌ Error: Template '{template}' not found at {template_dir}")
        return False

    # Output path
    plugin_name = f"{name.replace(' ', '')}.omnifocusjs"
    output_path = Path(output_dir) / plugin_name

    if output_path.exists():
        print(f"❌ Error: Plugin already exists at {output_path}")
        response = input("Overwrite? (yes/no): ")
        if response.lower() not in ['yes', 'y']:
            print("Cancelled.")
            return False
        shutil.rmtree(output_path)

    # Prepare template variables
    action_id = camel_case(name)
    short_label = name.split()[0] if ' ' in name else name
    variables = {
        'PLUGIN_NAME': name,
        'IDENTIFIER': identifier,
        'AUTHOR': author,
        'DESCRIPTION': description,
        'ACTION_ID': action_id,
        'ACTION_LABEL': f"Show {name}",
        'SHORT_LABEL': short_label,
        'MEDIUM_LABEL': f"Show {name}",
        'LONG_LABEL': f"Show {name} with OmniFocus Automation",
        'PALETTE_LABEL': name,
        'ICON': 'list.bullet',
    }

    print(f"\n🔨 Generating bundle plugin from template '{template}'...")
    print(f"   Plugin: {name}")
    print(f"   Identifier: {identifier}")
    print(f"   Output: {output_path}\n")

    # Copy template directory
    try:
        shutil.copytree(template_dir, output_path)
        print(f"✅ Copied template to {output_path}")
    except Exception as e:
        print(f"❌ Error copying template: {e}")
        return False

    # Process all files
    processed = 0
    js_files = []
    for root_dir, dirs, files in os.walk(output_path):
        # Skip README files (template documentation)
        dirs[:] = [d for d in dirs if d not in ['.git']]

        for file in files:
            if file == 'README.md':
                # Remove template README
                os.remove(Path(root_dir) / file)
                continue

            file_path = Path(root_dir) / file

            # Rename action.js to match ACTION_ID
            if file == 'action.js' and 'Resources' in root_dir and 'en.lproj' not in root_dir:
                new_name = f"{action_id}.js"
                new_path = Path(root_dir) / new_name
                os.rename(file_path, new_path)
                file_path = new_path
                print(f"  → Renamed action.js to {new_name}")

                # Also rename the .strings file if it exists
                old_strings = Path(root_dir) / 'en.lproj' / 'action.strings'
                if old_strings.exists():
                    new_strings = Path(root_dir) / 'en.lproj' / f'{action_id}.strings'
                    os.rename(old_strings, new_strings)
                    print(f"  → Renamed action.strings to {action_id}.strings")

            if process_file(file_path, variables):
                processed += 1

            if file_path.suffix == '.js':
                js_files.append(file_path)

    print(f"✅ Processed {processed} files")

    # Validate JavaScript files
    print(f"\n🔍 Validating JavaScript files...")
    errors = 0
    for js_file in js_files:
        if js_file.exists() and not validate_javascript(js_file):
            errors += 1

    if errors > 0:
        print(f"\n⚠️  Plugin generated with {errors} linting warning(s)")
        print(f"   Fix issues before installing to OmniFocus")
    else:
        print(f"\n🎉 Plugin generated successfully!")

    print(f"\n📦 Installation:")
    print(f"   cp -r {output_path} ~/Library/Application\\ Scripts/com.omnigroup.OmniFocus3/Plug-Ins/")
    print(f"\n🧪 Testing:")
    print(f"   1. Restart OmniFocus (if running)")
    print(f"   2. Go to Automation → {name}")
    print(f"   3. Select 'Show {name}' action")
    print(f"\n✏️  Customization:")
    print(f"   Edit: {output_path}/Resources/*.js")
    print(f"   Manifest: {output_path}/manifest.json")
    print(f"   Localization: {output_path}/Resources/en.lproj/*.strings")

    return True


def main():
    parser = argparse.ArgumentParser(
        description='Generate OmniFocus plugin from template',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog='''
Examples:
  # Simple solitary plugin (single file)
  %(prog)s --format solitary --name "My Quick Action"

  # Solitary plugin with Foundation Models / Apple Intelligence
  %(prog)s --format solitary-fm --name "AI Task Analyzer"

  # Solitary library plugin
  %(prog)s --format solitary-library --name "My Utils"

  # Bundle plugin (folder with localization)
  %(prog)s --format bundle --template query-simple --name "My Tasks"

  # Bundle with custom identifier
  %(prog)s --format bundle --template stats-overview --name "Overview" \\
    --identifier com.mycompany.overview --author "John Doe"

Available formats:
''' + '\n'.join(f"  {name}: {desc}" for name, desc in FORMATS.items()) + '''

Bundle templates:
''' + '\n'.join(f"  {name}: {desc}" for name, desc in BUNDLE_TEMPLATES.items())
    )

    parser.add_argument('--format', '-f',
                        choices=FORMATS.keys(),
                        required=True,
                        help='Plugin format type')
    parser.add_argument('--template', '-t',
                        choices=BUNDLE_TEMPLATES.keys(),
                        help='Bundle template (required for --format bundle)')
    parser.add_argument('--name', '-n',
                        required=True,
                        help='Plugin name (e.g., "My Plugin")')
    parser.add_argument('--identifier', '-i',
                        help='Reverse domain identifier (default: auto-generated)')
    parser.add_argument('--author', '-a',
                        help='Plugin author (default: "OmniFocus User")')
    parser.add_argument('--description', '-d',
                        help='Plugin description (default: auto-generated)')
    parser.add_argument('--output', '-o',
                        help='Output directory (default: assets/)')

    args = parser.parse_args()

    # Validate bundle requires template
    if args.format == 'bundle' and not args.template:
        parser.error("--template is required when --format is 'bundle'")

    # Generate based on format
    if args.format in ['solitary', 'solitary-fm', 'solitary-library']:
        success = generate_solitary_plugin(
            format_type=args.format,
            name=args.name,
            identifier=args.identifier,
            author=args.author,
            description=args.description,
            output_dir=args.output
        )
    else:  # bundle
        success = generate_bundle_plugin(
            template=args.template,
            name=args.name,
            identifier=args.identifier,
            author=args.author,
            description=args.description,
            output_dir=args.output
        )

    sys.exit(0 if success else 1)


if __name__ == '__main__':
    main()

File: skills/omnifocus-manager/scripts/generate_plugin.ts
#!/usr/bin/env node
/**
 * OmniFocus Plugin Generator (TypeScript)
 *
 * Generates OmniFocus plugins from TypeScript templates with integrated LSP validation.
 * Replaces the Python-based generator with TypeScript Compiler API validation.
 *
 * Usage:
 *   node generate_plugin.js --format solitary --name "My Plugin"
 *   node generate_plugin.js --format solitary-fm --name "AI Analyzer"
 *   node generate_plugin.js --format bundle --template query-simple --name "Tasks"
 *
 * @version 4.0.0
 * @author OmniFocus Manager Skill
 */

import * as ts from 'typescript';
import * as fs from 'fs/promises';
import * as fsSync from 'fs';
import * as path from 'path';
import { fileURLToPath } from 'url';

// ============================================================================
// Types
// ============================================================================

type PluginFormat = 'solitary' | 'solitary-fm' | 'solitary-library' | 'bundle';
type BundleTemplate = 'query-simple' | 'stats-overview';

interface PluginOptions {
  format: PluginFormat;
  template?: BundleTemplate;
  name: string;
  identifier?: string;
  author?: string;
  description?: string;
  outputDir?: string;
}

interface TemplateVariables {
  PLUGIN_NAME: string;
  IDENTIFIER: string;
  AUTHOR: string;
  DESCRIPTION: string;
  ACTION_ID: string;
  ACTION_LABEL: string;
  SHORT_LABEL: string;
  MEDIUM_LABEL: string;
  LONG_LABEL: string;
  PALETTE_LABEL: string;
  ICON: string;
}

interface ValidationResult {
  success: boolean;
  errors: Array<{
    line?: number;
    column?: number;
    message: string;
  }>;
}

interface GenerateResult {
  success: boolean;
  path?: string;
  errors?: string[];
}

// ============================================================================
// Constants
// ============================================================================

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const FORMATS: Record<PluginFormat, string> = {
  'solitary': 'Single-file action plugin (.omnijs)',
  'solitary-fm': 'Single-file action with Foundation Models (.omnijs)',
  'solitary-library': 'Single-file library plugin (.omnijs)',
  'bundle': 'Bundle plugin with localization (.omnifocusjs folder)',
};

const BUNDLE_TEMPLATES: Record<BundleTemplate, string> = {
  'query-simple': 'Simple query plugin with Alert display',
  'stats-overview': 'Statistics overview dashboard',
};

// ============================================================================
// Utility Functions
// ============================================================================

/**
 * Find omnifocus-manager skill root directory
 */
function findSkillRoot(): string | null {
  let current = path.resolve(__dirname);
  while (current !== path.dirname(current)) {
    const skillFile = path.join(current, 'SKILL.md');
    if (fsSync.existsSync(skillFile) && path.basename(current) === 'omnifocus-manager') {
      return current;
    }
    current = path.dirname(current);
  }
  return null;
}

/**
 * Convert text to camelCase
 */
function camelCase(text: string): string {
  const words = text.replace(/-/g, ' ').replace(/_/g, ' ').split(/\s+/);
  if (words.length === 0) return '';
  return words[0]!.toLowerCase() + words.slice(1).map(w => w.charAt(0).toUpperCase() + w.slice(1)).join('');
}

/**
 * Replace template variables in content
 */
function substituteVariables(content: string, variables: TemplateVariables): string {
  let result = content;
  for (const [key, value] of Object.entries(variables)) {
    const placeholder = `{{${key}}}`;
    result = result.replace(new RegExp(placeholder.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'), 'g'), value);
  }
  return result;
}

/**
 * Prepare template variables from options
 */
function prepareVariables(options: PluginOptions): TemplateVariables {
  const identifier = options.identifier || `com.user.omnifocus.${options.name.toLowerCase().replace(/\s+/g, '')}`;
  const author = options.author || 'OmniFocus User';
  const description = options.description || `${options.name} plugin for OmniFocus`;
  const actionId = camelCase(options.name);
  const shortLabel = options.name.split(/\s+/)[0] || options.name;

  const icon = options.format === 'solitary-fm' ? 'apple.intelligence' : 'list.bullet';

  return {
    PLUGIN_NAME: options.name,
    IDENTIFIER: identifier,
    AUTHOR: author,
    DESCRIPTION: description,
    ACTION_ID: actionId,
    ACTION_LABEL: options.name,
    SHORT_LABEL: shortLabel,
    MEDIUM_LABEL: options.name,
    LONG_LABEL: `${options.name} with OmniFocus Automation`,
    PALETTE_LABEL: options.name,
    ICON: icon,
  };
}

// ============================================================================
// Plugin Generator Class
// ============================================================================

class PluginGenerator {
  private skillRoot: string;
  private dtsPath: string;
  private dtsExtensionsPath: string;

  constructor() {
    const root = findSkillRoot();
    if (!root) {
      throw new Error('Could not find omnifocus-manager skill root directory');
    }

    this.skillRoot = root;
    this.dtsPath = path.join(root, 'typescript', 'omnifocus.d.ts');
    this.dtsExtensionsPath = path.join(root, 'typescript', 'omnifocus-extensions.d.ts');
  }

  /**
   * Generate plugin from options
   */
  async generate(options: PluginOptions): Promise<GenerateResult> {
    try {
      console.log(`\n🔨 Generating ${options.format} plugin...`);
      console.log(`   Plugin: ${options.name}`);
      console.log(`   Format: ${FORMATS[options.format]}\n`);

      // 1. Load template
      const template = await this.loadTemplate(options);

      // 2. Prepare variables and substitute
      const variables = prepareVariables(options);
      const code = substituteVariables(template, variables);

      // 3. CRITICAL: Validate with TypeScript compiler
      console.log('🔍 Validating TypeScript...');
      const validation = await this.validateTypeScript(code, variables.ACTION_ID);

      if (!validation.success) {
        console.log('\n❌ TypeScript validation failed:');
        validation.errors.forEach((err, i) => {
          const location = err.line !== undefined ? `Line ${err.line + 1}${err.column !== undefined ? `:${err.column + 1}` : ''}` : '';
          console.log(`   ${i + 1}. ${location ? location + ' - ' : ''}${err.message}`);
        });
        return { success: false, errors: validation.errors.map(e => e.message) };
      }

      console.log('   ✅ TypeScript validation passed\n');

      // 4. Write .ts file
      const tsFilePath = await this.writeTsFile(code, options);

      // 5. Auto-rename to .omnijs/.omnifocusjs
      const deployPath = await this.renameForDeployment(tsFilePath, options.format);

      console.log('🎉 Plugin generated successfully!\n');
      console.log('📦 Installation:');
      if (options.format === 'bundle') {
        console.log(`   cp -r ${deployPath} ~/Library/Application\\ Scripts/com.omnigroup.OmniFocus3/Plug-Ins/`);
      } else {
        console.log(`   cp ${deployPath} ~/Library/Application\\ Scripts/com.omnigroup.OmniFocus3/Plug-Ins/`);
      }
      console.log('\n🧪 Testing:');
      console.log('   1. Restart OmniFocus (if running)');
      console.log('   2. Go to Automation menu');
      console.log(`   3. Find '${options.name}'\n`);

      return { success: true, path: deployPath };
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      console.error(`\n❌ Error: ${message}\n`);
      return { success: false, errors: [message] };
    }
  }

  /**
   * Load template based on format
   */
  private async loadTemplate(options: PluginOptions): Promise<string> {
    let templatePath: string;

    if (options.format === 'bundle') {
      if (!options.template) {
        throw new Error('Bundle format requires --template option');
      }
      templatePath = path.join(
        this.skillRoot,
        'assets',
        'plugin-templates',
        options.template,
        'Resources',
        'action.js'
      );
    } else {
      const formatMap: Record<Exclude<PluginFormat, 'bundle'>, string> = {
        'solitary': 'solitary-action/action.ts.template',
        'solitary-fm': 'solitary-action-fm/action.ts.template',
        'solitary-library': 'solitary-library/library.ts.template',
      };

      templatePath = path.join(
        this.skillRoot,
        'assets',
        'plugin-templates',
        formatMap[options.format]
      );
    }

    try {
      return await fs.readFile(templatePath, 'utf-8');
    } catch (error) {
      throw new Error(`Template not found: ${templatePath}`);
    }
  }

  /**
   * Validate TypeScript code using TypeScript Compiler API
   *
   * Note: This is a simplified validation that checks for OmniFocus API usage.
   * We skip full TypeScript lib validation and focus on OmniFocus-specific errors.
   */
  private async validateTypeScript(code: string, fileName: string): Promise<ValidationResult> {
    try {
      // Read type definitions
      const omnifocusTypes = await fs.readFile(this.dtsPath, 'utf-8');
      const extensionTypes = await fs.readFile(this.dtsExtensionsPath, 'utf-8');

      // Prepend type definitions to code for validation
      // This is a simpler approach than setting up a full compiler host
      const fullCode = `
// OmniFocus Type Definitions
${omnifocusTypes}

// OmniFocus Extensions
${extensionTypes}

// Plugin Code
${code}
`;

      // Create source file for syntax check
      const sourceFile = ts.createSourceFile(
        `${fileName}.ts`,
        fullCode,
        ts.ScriptTarget.ES2022,
        true,
        ts.ScriptKind.TS
      );

      // Check for syntax errors by examining the source file
      // If TypeScript can parse it successfully, the plugin code is syntactically valid
      const hasErrors = this.hasSyntaxErrors(sourceFile);

      if (hasErrors) {
        return {
          success: false,
          errors: [{ message: 'TypeScript syntax error detected. Please check the template or generated code.' }]
        };
      }

      // Code is syntactically valid
      // TypeScript LSP will provide additional semantic validation during development
      return { success: true, errors: [] };

    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      return {
        success: false,
        errors: [{ message: `TypeScript validation error: ${message}` }]
      };
    }
  }

  /**
   * Check if source file has syntax errors
   */
  private hasSyntaxErrors(node: ts.Node): boolean {
    // Check if current node has syntax errors
    if (node.kind === ts.SyntaxKind.Unknown) {
      return true;
    }

    // Recursively check children
    return ts.forEachChild(node, child => this.hasSyntaxErrors(child)) || false;
  }

  /**
   * Write TypeScript file
   */
  private async writeTsFile(
    code: string,
    options: PluginOptions
  ): Promise<string> {
    const outputDir = options.outputDir || path.join(this.skillRoot, 'assets');
    const fileName = `${options.name.replace(/\s+/g, '')}.ts`;
    const filePath = path.join(outputDir, fileName);

    await fs.writeFile(filePath, code, 'utf-8');
    console.log(`✅ Created ${filePath}`);

    return filePath;
  }

  /**
   * Rename .ts file to deployment format (.omnijs or .omnifocusjs)
   */
  private async renameForDeployment(tsFilePath: string, format: PluginFormat): Promise<string> {
    const ext = format === 'bundle' ? '.omnifocusjs' : '.omnijs';
    const deployPath = tsFilePath.replace(/\.ts$/, ext);

    await fs.rename(tsFilePath, deployPath);
    console.log(`✅ Renamed to ${deployPath}`);

    return deployPath;
  }
}

// ============================================================================
// CLI Argument Parsing
// ============================================================================

function parseArgs(args: string[]): PluginOptions | null {
  const options: Partial<PluginOptions> = {};

  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    const next = args[i + 1];

    switch (arg) {
      case '--format':
      case '-f':
        if (next && (next in FORMATS)) {
          options.format = next as PluginFormat;
          i++;
        }
        break;

      case '--template':
      case '-t':
        if (next && (next in BUNDLE_TEMPLATES)) {
          options.template = next as BundleTemplate;
          i++;
        }
        break;

      case '--name':
      case '-n':
        if (next) {
          options.name = next;
          i++;
        }
        break;

      case '--identifier':
      case '-i':
        if (next) {
          options.identifier = next;
          i++;
        }
        break;

      case '--author':
      case '-a':
        if (next) {
          options.author = next;
          i++;
        }
        break;

      case '--description':
      case '-d':
        if (next) {
          options.description = next;
          i++;
        }
        break;

      case '--output':
      case '-o':
        if (next) {
          options.outputDir = next;
          i++;
        }
        break;

      case '--help':
      case '-h':
        printUsage();
        return null;
    }
  }

  // Validate required options
  if (!options.format) {
    console.error('❌ Error: --format is required\n');
    printUsage();
    return null;
  }

  if (!options.name) {
    console.error('❌ Error: --name is required\n');
    printUsage();
    return null;
  }

  if (options.format === 'bundle' && !options.template) {
    console.error('❌ Error: --template is required when --format is bundle\n');
    printUsage();
    return null;
  }

  return options as PluginOptions;
}

function printUsage(): void {
  console.log(`
OmniFocus Plugin Generator (TypeScript)

Usage:
  node generate_plugin.js [options]

Required Options:
  --format, -f <format>         Plugin format: ${Object.keys(FORMATS).join(', ')}
  --name, -n <name>             Plugin name (e.g., "My Plugin")

Optional Options:
  --template, -t <template>     Bundle template: ${Object.keys(BUNDLE_TEMPLATES).join(', ')} (required for bundle format)
  --identifier, -i <id>         Reverse domain identifier (default: auto-generated)
  --author, -a <author>         Plugin author (default: "OmniFocus User")
  --description, -d <desc>      Plugin description (default: auto-generated)
  --output, -o <dir>            Output directory (default: assets/)
  --help, -h                    Show this help message

Examples:
  # Simple solitary plugin
  node generate_plugin.js --format solitary --name "My Quick Action"

  # Solitary plugin with Foundation Models
  node generate_plugin.js --format solitary-fm --name "AI Task Analyzer"

  # Bundle plugin
  node generate_plugin.js --format bundle --template query-simple --name "My Tasks"

Available Formats:
${Object.entries(FORMATS).map(([name, desc]) => `  ${name.padEnd(20)} ${desc}`).join('\n')}

Bundle Templates:
${Object.entries(BUNDLE_TEMPLATES).map(([name, desc]) => `  ${name.padEnd(20)} ${desc}`).join('\n')}
`);
}

// ============================================================================
// Main
// ============================================================================

async function main(): Promise<void> {
  const args = process.argv.slice(2);

  if (args.length === 0 || args.includes('--help') || args.includes('-h')) {
    printUsage();
    process.exit(0);
  }

  const options = parseArgs(args);
  if (!options) {
    process.exit(1);
  }

  const generator = new PluginGenerator();
  const result = await generator.generate(options);

  process.exit(result.success ? 0 : 1);
}

// Run if called directly
if (import.meta.url === `file://${__filename}`) {
  main().catch(error => {
    console.error('Fatal error:', error);
    process.exit(1);
  });
}

export { PluginGenerator, type PluginOptions, type GenerateResult };
